[JSC] Add core semantics of import defer

takikawa opened the pull request:
#### 76b9ea94098a0a52860f9c98d8feec6f0f01f7a8
<pre>
[JSC] Add core semantics of import defer

<a href="https://bugs.webkit.org/show_bug.cgi?id=286272">https://bugs.webkit.org/show_bug.cgi?id=286272</a>

Reviewed by NOBODY (OOPS!).

Implements the semantics of `import defer` along with deferred namespace
objects. The proposal semantics can be found at:

  <a href="https://github.com/tc39/proposal-defer-import-eval/">https://github.com/tc39/proposal-defer-import-eval/</a>

This doesn&apos;t include the dynamic form `import.defer()` yet. There are a few
limitations of the implementation that mostly stem from other bugs in the
module loader. The following test262 tests do not pass yet and have
test expectations set:

  - import-defer/errors/module-throws/trigger-evaluation.js
  - import-defer/errors/module-throws/defer-import-after-evaluation.js
  - import-defer/errors/module-throws/third-party-evaluation-after-defer-import.js
    (because JSC does not store and re-throw module eval errors)
  - import-defer/errors/get-other-while-evaluating-async
  - import-defer/errors/get-other-while-dep-evaluating-async
    (in debug mode, due to a assertion that can be triggered already in JSC)
  - import-defer/evaluation-top-level-await/flattening-order
    (due to async module behavior in JSC)
  - import-defer/deferred-namespace-object/exotic-object-behavior.js
  - import-defer/errors/get-self-while-defer-evaluating/main.js
  - import-defer/errors/resolution-error/import-defer-of-missing-module-fails.js:
    (These tests fail temporarily due to upstream test bugs,
     should get fixed after test262 gets updated)

Many of these issues will be fixed by the general module refactoring for
the following bug: <a href="https://bugs.webkit.org/show_bug.cgi?id=242740">https://bugs.webkit.org/show_bug.cgi?id=242740</a>

* JSTests/modules/import-defer-1-1-cycle.js: Added.
* JSTests/modules/import-defer-1-1-cycle/a.js: Added.
* JSTests/modules/import-defer-1-1-cycle/b.js: Added.
* JSTests/modules/import-defer-1-1-cycle/c.js: Added.
* JSTests/modules/import-defer-1-1-cycle/d.js: Added.
* JSTests/modules/import-defer-1-1-cycle/e.js: Added.
* JSTests/modules/import-defer-1-cycle.js: Added.
* JSTests/modules/import-defer-1-cycle/a.js: Added.
* JSTests/modules/import-defer-1-cycle/c.js: Added.
* JSTests/modules/import-defer-1-cycle/e.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/a.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/b.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/c.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/d.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/e.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/b.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/d.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/a.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/c.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/e.js: Added.
* JSTests/modules/import-defer-2-force-evaluation.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/b.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/d.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/a.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/b.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/c.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/d.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/e.js: Added.
* JSTests/modules/import-defer-3-force-evaluation.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/a.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/c.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/e.js: Added.
* JSTests/modules/import-defer-4-both-imports.js: Added.
* JSTests/modules/import-defer-4-both-imports/a.js: Added.
* JSTests/modules/import-defer-4-both-imports/b.js: Added.
* JSTests/modules/import-defer-4-both-imports/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/f.js: Added.
* JSTests/modules/import-defer-5-multiple-async.js: Added.
* JSTests/modules/import-defer-5-multiple-async/a.js: Added.
* JSTests/modules/import-defer-5-multiple-async/b.js: Added.
* JSTests/modules/import-defer-5-multiple-async/c.js: Added.
* JSTests/modules/import-defer-5-multiple-async/d.js: Added.
* JSTests/modules/import-defer-5-multiple-async/e.js: Added.
* JSTests/modules/import-defer-5-multiple-async/f.js: Added.
* JSTests/modules/import-defer-6-force-eval-order.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/a.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/b.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/c.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/d.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/e.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/f.js: Added.
* JSTests/modules/import-defer-basic.js: Added.
* JSTests/modules/import-defer-basic/1.js: Added.
(export.foo):
* JSTests/modules/import-defer-basic/2.js: Added.
* JSTests/modules/import-defer-default.js: Added.
* JSTests/modules/import-defer-default/foo.js: Added.
* JSTests/modules/import-defer-defineownproperty.js: Added.
* JSTests/modules/import-defer-deleteproperty.js: Added.
* JSTests/modules/import-defer-getownproperty.js: Added.
* JSTests/modules/import-defer-hasproperty.js: Added.
* JSTests/modules/import-defer-ownpropertykeys.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-ready-for-sync-execution-cycle.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-then.js: Added.
* JSTests/modules/import-defer-then/export-then.js: Added.
* JSTests/modules/import-defer-tla.js: Added.
* JSTests/modules/import-defer-tla/b.js: Added.
* JSTests/modules/import-defer-tla/c.js: Added.
* JSTests/modules/import-defer-tla/d.js: Added.
* JSTests/modules/import-defer-tla/e.js: Added.
* JSTests/modules/import-defer-tla/f.js: Added.
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(linkTimeConstant.newRegistryEntry):
(visibility.PrivateRecursive.requestSatisfyUtil):
(visibility.PrivateRecursive.link):
(visibility.PrivateRecursive.moduleEvaluation):
(visibility.PrivateRecursive.async asyncModuleEvaluation):
(visibility.PrivateRecursive.async asyncModuleDeferredEvaluation):
(visibility.PrivateRecursive.async requestImportModule):
(visibility.PrivateRecursive.dependencyKeysIfEvaluated):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/parser/ModuleAnalyzer.cpp:
(JSC::ModuleAnalyzer::appendRequestedModule):
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/JavaScriptCore/parser/NodesAnalyzeModule.cpp:
(JSC::ImportDeclarationNode::analyzeModule):
(JSC::ExportAllDeclarationNode::analyzeModule):
(JSC::ExportNamedDeclarationNode::analyzeModule):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::visitChildrenImpl):
(JSC::AbstractModuleRecord::appendRequestedModule):
(JSC::AbstractModuleRecord::getModuleNamespace):
(JSC::AbstractModuleRecord::dump):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSModuleLoader::finishCreation):
(JSC::JSModuleLoader::getModuleNamespaceObject):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSModuleLoader.h:
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
(JSC::JSModuleNamespaceObject::JSModuleNamespaceObject):
(JSC::JSModuleNamespaceObject::finishCreation):
(JSC::JSModuleNamespaceObject::ensureDeferredNamespaceEvaluation):
(JSC::JSModuleNamespaceObject::isSymbolLikeNamespaceKey):
(JSC::JSModuleNamespaceObject::getOwnPropertySlotCommon):
(JSC::JSModuleNamespaceObject::moduleExportsListContains):
(JSC::JSModuleNamespaceObject::deleteProperty):
(JSC::JSModuleNamespaceObject::deletePropertyByIndex):
(JSC::JSModuleNamespaceObject::getOwnPropertyNames):
(JSC::JSModuleNamespaceObject::defineOwnProperty):
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.h:
* Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
(JSC::JSModuleRecord::instantiateDeclarations):
(JSC::JSModuleRecord::evaluate):
(JSC::JSModuleRecord::readyForSyncExecution):
* Source/JavaScriptCore/runtime/JSModuleRecord.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::tryCreate):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76b9ea94098a0a52860f9c98d8feec6f0f01f7a8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13772 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108941 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54402 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23848 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78832 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93592 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18313 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11641 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53778 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96425 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88083 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111330 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102361 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22788 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31267 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89793 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87481 "Found 101 new API test failures: /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/sequence, /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins, /TestWebKit:WebKit.GeolocationBasicWithHighAccuracy, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/cancel-sequence, /TestWebKit:WebKit.ForceRepaint, /TestWebKit:WebKit.PendingAPIRequestURL, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/mediaKeySystem-permission-requests, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit2.ProvisionalURLAfterWillSendRequestCallback, /WebKitGTK/TestDownloads:/webkit/Downloads/contex-menu-download-actions ... (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32361 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10088 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25263 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30834 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36137 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125994 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30628 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34881 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

takikawa committed:
[JSC] Add core semantics of import defer

https://bugs.webkit.org/show_bug.cgi?id=286272

Reviewed by NOBODY (OOPS!).

Implements the semantics of `import defer` along with deferred namespace
objects. The proposal semantics can be found at:

  https://github.com/tc39/proposal-defer-import-eval/

This doesn't include the dynamic form `import.defer()` yet. There are a few
limitations of the implementation that mostly stem from other bugs in the
module loader. The following test262 tests do not pass yet and have
test expectations set:

  - import-defer/errors/module-throws/trigger-evaluation.js
  - import-defer/errors/module-throws/defer-import-after-evaluation.js
  - import-defer/errors/module-throws/third-party-evaluation-after-defer-import.js
    (because JSC does not store and re-throw module eval errors)
  - import-defer/errors/get-other-while-evaluating-async
  - import-defer/errors/get-other-while-dep-evaluating-async
    (in debug mode, due to a assertion that can be triggered already in JSC)
  - import-defer/evaluation-top-level-await/flattening-order
    (due to async module behavior in JSC)
  - import-defer/deferred-namespace-object/exotic-object-behavior.js
  - import-defer/errors/get-self-while-defer-evaluating/main.js
  - import-defer/errors/resolution-error/import-defer-of-missing-module-fails.js:
    (These tests fail temporarily due to upstream test bugs,
     should get fixed after test262 gets updated)

Many of these issues will be fixed by the general module refactoring for
the following bug: https://bugs.webkit.org/show_bug.cgi?id=242740

* JSTests/modules/import-defer-1-1-cycle.js: Added.
* JSTests/modules/import-defer-1-1-cycle/a.js: Added.
* JSTests/modules/import-defer-1-1-cycle/b.js: Added.
* JSTests/modules/import-defer-1-1-cycle/c.js: Added.
* JSTests/modules/import-defer-1-1-cycle/d.js: Added.
* JSTests/modules/import-defer-1-1-cycle/e.js: Added.
* JSTests/modules/import-defer-1-cycle.js: Added.
* JSTests/modules/import-defer-1-cycle/a.js: Added.
* JSTests/modules/import-defer-1-cycle/c.js: Added.
* JSTests/modules/import-defer-1-cycle/e.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/a.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/b.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/c.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/d.js: Added.
* JSTests/modules/import-defer-2-1-deferred-cycle/e.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/b.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/d.js: Added.
* JSTests/modules/import-defer-2-1-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/a.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/c.js: Added.
* JSTests/modules/import-defer-2-deferred-cycle/e.js: Added.
* JSTests/modules/import-defer-2-force-evaluation.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-2-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/b.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/d.js: Added.
* JSTests/modules/import-defer-3-1-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/a.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/b.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/c.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/d.js: Added.
* JSTests/modules/import-defer-3-1-indirect-cycle/e.js: Added.
* JSTests/modules/import-defer-3-force-evaluation.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/a.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/c.js: Added.
* JSTests/modules/import-defer-3-force-evaluation/e.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/a.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/c.js: Added.
* JSTests/modules/import-defer-3-indirect-cycle/e.js: Added.
* JSTests/modules/import-defer-4-both-imports.js: Added.
* JSTests/modules/import-defer-4-both-imports/a.js: Added.
* JSTests/modules/import-defer-4-both-imports/b.js: Added.
* JSTests/modules/import-defer-4-both-imports/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-a-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d-b-a/f.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/a.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/b.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/c.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/d.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/e.js: Added.
* JSTests/modules/import-defer-5-force-evaluation-e-d/f.js: Added.
* JSTests/modules/import-defer-5-multiple-async.js: Added.
* JSTests/modules/import-defer-5-multiple-async/a.js: Added.
* JSTests/modules/import-defer-5-multiple-async/b.js: Added.
* JSTests/modules/import-defer-5-multiple-async/c.js: Added.
* JSTests/modules/import-defer-5-multiple-async/d.js: Added.
* JSTests/modules/import-defer-5-multiple-async/e.js: Added.
* JSTests/modules/import-defer-5-multiple-async/f.js: Added.
* JSTests/modules/import-defer-6-force-eval-order.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/a.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/b.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/c.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/d.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/e.js: Added.
* JSTests/modules/import-defer-6-force-eval-order/f.js: Added.
* JSTests/modules/import-defer-basic.js: Added.
* JSTests/modules/import-defer-basic/1.js: Added.
(export.foo):
* JSTests/modules/import-defer-basic/2.js: Added.
* JSTests/modules/import-defer-default.js: Added.
* JSTests/modules/import-defer-default/foo.js: Added.
* JSTests/modules/import-defer-defineownproperty.js: Added.
* JSTests/modules/import-defer-deleteproperty.js: Added.
* JSTests/modules/import-defer-getownproperty.js: Added.
* JSTests/modules/import-defer-hasproperty.js: Added.
* JSTests/modules/import-defer-ownpropertykeys.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-delete/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle-keys/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-ready-for-sync-execution-cycle.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/a.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/b.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/c.js: Added.
* JSTests/modules/import-defer-ready-for-sync-execution-cycle/d.js: Added.
(shouldThrow):
* JSTests/modules/import-defer-then.js: Added.
* JSTests/modules/import-defer-then/export-then.js: Added.
* JSTests/modules/import-defer-tla.js: Added.
* JSTests/modules/import-defer-tla/b.js: Added.
* JSTests/modules/import-defer-tla/c.js: Added.
* JSTests/modules/import-defer-tla/d.js: Added.
* JSTests/modules/import-defer-tla/e.js: Added.
* JSTests/modules/import-defer-tla/f.js: Added.
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(linkTimeConstant.newRegistryEntry):
(visibility.PrivateRecursive.requestSatisfyUtil):
(visibility.PrivateRecursive.link):
(visibility.PrivateRecursive.moduleEvaluation):
(visibility.PrivateRecursive.async asyncModuleEvaluation):
(visibility.PrivateRecursive.async asyncModuleDeferredEvaluation):
(visibility.PrivateRecursive.async requestImportModule):
(visibility.PrivateRecursive.dependencyKeysIfEvaluated):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/parser/ModuleAnalyzer.cpp:
(JSC::ModuleAnalyzer::appendRequestedModule):
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/JavaScriptCore/parser/NodesAnalyzeModule.cpp:
(JSC::ImportDeclarationNode::analyzeModule):
(JSC::ExportAllDeclarationNode::analyzeModule):
(JSC::ExportNamedDeclarationNode::analyzeModule):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::visitChildrenImpl):
(JSC::AbstractModuleRecord::appendRequestedModule):
(JSC::AbstractModuleRecord::getModuleNamespace):
(JSC::AbstractModuleRecord::dump):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSModuleLoader::finishCreation):
(JSC::JSModuleLoader::getModuleNamespaceObject):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSModuleLoader.h:
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
(JSC::JSModuleNamespaceObject::JSModuleNamespaceObject):
(JSC::JSModuleNamespaceObject::finishCreation):
(JSC::JSModuleNamespaceObject::ensureDeferredNamespaceEvaluation):
(JSC::JSModuleNamespaceObject::isSymbolLikeNamespaceKey):
(JSC::JSModuleNamespaceObject::getOwnPropertySlotCommon):
(JSC::JSModuleNamespaceObject::moduleExportsListContains):
(JSC::JSModuleNamespaceObject::deleteProperty):
(JSC::JSModuleNamespaceObject::deletePropertyByIndex):
(JSC::JSModuleNamespaceObject::getOwnPropertyNames):
(JSC::JSModuleNamespaceObject::defineOwnProperty):
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.h:
* Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
(JSC::JSModuleRecord::instantiateDeclarations):
(JSC::JSModuleRecord::evaluate):
(JSC::JSModuleRecord::readyForSyncExecution):
* Source/JavaScriptCore/runtime/JSModuleRecord.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::tryCreate):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/990aba9095bb4c6ec91c88191edf02ac27f63468. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7c6ef147df81e6909f7ab7993f3231695c1fa761. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

takikawa commented:
In the latest version I fixed two bugs, one in which cyclic module graphs would non-terminate when the cycles were between deferred/non-evaluated modules, and one due to the "ready for sync execution" check. Fixed API tests as well. Also refactored some of the module loader JS code.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d4a1b37985319c2ca148c6cb453e54387a2329be. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

Constellation commented on Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
```diff
@@ -205,6 +208,7 @@ class AbstractModuleRecord : public JSInternalFieldObjectImpl<2> {
     WriteBarrier<JSMap> m_dependenciesMap;
     
     WriteBarrier<JSModuleNamespaceObject> m_moduleNamespaceObject;
+    WriteBarrier<JSModuleNamespaceObject> m_moduleDeferredNamespaceObject;
```
This new field is not visited by GC.

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -117,6 +119,31 @@ static JSValue getValue(JSModuleEnvironment* environment, PropertyName localName
+    }
+
+    // This calls linkAndEvaluateModule instead of just evaluate in order to
+    // trigger evaluation for dependency modules as well.
```
We should do `scope.release()`.

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -117,6 +119,31 @@ static JSValue getValue(JSModuleEnvironment* environment, PropertyName localName
+    // This calls linkAndEvaluateModule instead of just evaluate in order to
+    // trigger evaluation for dependency modules as well.
+    globalObject->moduleLoader()->linkAndEvaluateModule(globalObject, identifierToJSValue(vm, m_moduleRecord->moduleKey()), jsUndefined());
+    RETURN_IF_EXCEPTION(scope, void());
```
This is not necessary.

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +244,38 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
+        ensureDeferredNamespaceEvaluation(globalObject);
+        RETURN_IF_EXCEPTION(scope, false);
+    }
+    RELEASE_AND_RETURN(scope, !!m_exports.contains(name));
```
Why is it `RELEASE_AND_RETURN`? It should be `return m_exports.contains(name);`

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +244,38 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
 bool JSModuleNamespaceObject::deletePropertyByIndex(JSCell* cell, JSGlobalObject* globalObject, unsigned propertyName)
 {
     VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
```
Why do we need this scope?

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -117,6 +119,31 @@ static JSValue getValue(JSModuleEnvironment* environment, PropertyName localName
+    RETURN_IF_EXCEPTION(scope, void());
+}
+
+bool JSModuleNamespaceObject::isSymbolLikeNamespaceKey(JSGlobalObject* globalObject, PropertyName propertyName)
```
Why not taking VM& instead of `JSGlobalObject*`?

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +244,38 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
+
 bool JSModuleNamespaceObject::deleteProperty(JSCell* cell, JSGlobalObject* globalObject, PropertyName propertyName, DeletePropertySlot& slot)
 {
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
```
Why do we need this scope?

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +244,38 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
         return Base::deleteProperty(thisObject, globalObject, propertyName, slot);
 
-    return !thisObject->m_exports.contains(propertyName.uid());
+    RELEASE_AND_RETURN(scope, !thisObject->moduleExportsListContains(globalObject, propertyName.uid()));
```
Why do we need to use `RELEASE_AND_RETURN` when it is just returning?

Constellation commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +244,38 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
+
     JSModuleNamespaceObject* thisObject = jsCast<JSModuleNamespaceObject*>(cell);
-    return !thisObject->m_exports.contains(Identifier::from(vm, propertyName).impl());
+    RELEASE_AND_RETURN(scope, !thisObject->moduleExportsListContains(globalObject, Identifier::from(vm, propertyName).impl()));
```
Ditto.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66474d3cda71b75a9b751892bc58fc903b7aacde. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c75cedc678658f86209ea494167a975520d503b2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

takikawa commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -212,21 +247,36 @@ bool JSModuleNamespaceObject::putByIndex(JSCell*, JSGlobalObject* globalObject,
     return false;
 }
 
+bool JSModuleNamespaceObject::moduleExportsListContains(JSGlobalObject* globalObject, const RefPtr<UniquedStringImpl> name)
```
Adjusted this signature (use `RefPtr`) due to Safe C++ checks

takikawa commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -117,6 +119,34 @@ static JSValue getValue(JSModuleEnvironment* environment, PropertyName localName
+    ASSERT(m_isDeferred);
+    if (JSModuleRecord* jsModule = jsDynamicCast<JSModuleRecord*>(moduleRecord()); jsModule) {
+        bool readyForSyncExecution = jsModule->readyForSyncExecution(globalObject);
+        RETURN_IF_EXCEPTION(scope, void());
```
While fixing the throw scope handling, I found that there's a check needed here (and in `readyForSyncExecution` as well). I believe the exception handling should be correct now as the module tests run fine in debug mode.

nicolo-ribaudo commented:
Het fyi, the proposal reached stage 3 at this week's TC39 meeting :)

takikawa commented on Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
```diff
@@ -262,9 +262,52 @@ JSValue JSModuleRecord::evaluate(JSGlobalObject* globalObject, JSValue sentValue
     ModuleProgramExecutable* executable = m_moduleProgramExecutable.get();
     JSValue resultOrAwaitedValue = vm.interpreter.executeModuleProgram(this, executable, globalObject, moduleEnvironment(), sentValue, resumeMode);
-    if (JSValue state = internalField(Field::State).get(); !state.isNumber() || state.asNumber() == static_cast<unsigned>(State::Executing))
+    if (JSValue state = internalField(Field::State).get(); !state.isNumber() || state.asInt32() == static_cast<int>(State::Executing))
```
BTW, I forgot to note this earlier (I can put it in the commit message) but this change is here to fix what I think is a bug in which `m_moduleProgramExecutable` never actually got cleared because this state check wouldn't be true.

justinmichaud replied:
Can we split this into a separate patch?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bb5408727f6247d0c60d694efc4d71bdac7a6b61. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

justinmichaud commented on Source/JavaScriptCore/builtins/ModuleLoader.js:
```diff
@@ -101,6 +101,8 @@ function newRegistryEntry(key)
         evaluated: false,
         then: @undefined,
         isAsync: false,
+        hasTLA: false,
```
Can we avoid acronyms?

justinmichaud commented on Source/JavaScriptCore/builtins/ModuleLoader.js:
```diff
@@ -209,13 +211,15 @@ function requestInstantiate(entry, parameters, fetcher)
         var moduleRecord = await this.parseModule(key, source);
         var dependenciesMap = moduleRecord.dependenciesMap;
         var requestedModules = this.requestedModules(moduleRecord);
+        var depPhases = this.requestedModulePhases(moduleRecord);
```
ditto. We say dependencies above

justinmichaud commented on Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
```diff
@@ -320,7 +323,15 @@ JSModuleNamespaceObject* JSModuleLoader::getModuleNamespaceObject(JSGlobalObject
+        throwTypeError(globalObject, scope, "expected a module phase value"_s);
+        return nullptr;
+    }
+    if (static_cast<Phase>(phaseValue.asUInt32()) == Phase::Defer)
```
Just a general question, is it normal for these to be numbers? It's not necessarily a problem, I just wish the JSValue was more expressive somehow

takikawa replied:
I believe so, that's how it's done for other intrinsic values listed in BytecodeIntrinsicRegistry.cpp for example. It's true that it's a bit error-prone if you read the JSValue incorrectly (this PR fixes an existing bug related to that). If there's a better way I'm happy to change it though.

justinmichaud commented on Source/JavaScriptCore/runtime/JSModuleNamespaceObject.cpp:
```diff
@@ -117,6 +119,34 @@ static JSValue getValue(JSModuleEnvironment* environment, PropertyName localName
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    ASSERT(Options::useImportDefer());
```
Release assert for now? Let's make sure all of these cases are super carefully guarded

justinmichaud reviewed (commented):
Overall, LGTM. I would like to see some more tests for the different places where we can throw an exception (basically, one for every place where we RETURN_IF_EXCEPTION). 

Some tests for cycles would be nice, even if they are skipped for now.

I am not super familiar with this code, so I don't feel comfortable r+ing.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ce53daf292d0bd083fe8acafd5ca877095422d7f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453

takikawa commented:
Thanks for taking a look!

> Overall, LGTM. I would like to see some more tests for the different places where we can throw an exception (basically, one for every place where we RETURN_IF_EXCEPTION).

I added a few more tests for the `readyForSyncExecution` check to make sure it's being checked for different operations on namespaces. There are some other theoretical exception paths which I think are difficult to trigger with an actual test (e.g., having `hostResolveImportModule` fail in a `readyForSyncExecution` call).

> Some tests for cycles would be nice, even if they are skipped for now.

Some of the test262 tests related to cycles are indeed skipped, but there should already be tests for other kinds of cycles in the module tests (e.g., `JSTests/module/import-defer-1-1-cycle.js`, etc).

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/76b9ea94098a0a52860f9c98d8feec6f0f01f7a8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/40453
